### PR TITLE
Add security automation lab type

### PR DESCRIPTION
**Pull request by maxamillion**
Signed-off-by: Adam Miller <admiller@redhat.com>

##### SUMMARY

This is still a work in progress but major building blocks are in
this pull request which allow the lab environment to be iterated on.

This PR includes Splunk Enterprise and snort setup and a RHEL
machine as a placeholder for the Checkpoint NGFW in the provision
process.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

security_automation

### CODE REVIEW

**Comment by maxamillion** [comment]
cc @IPvSean @liquidat

File: provisioner/group_vars/all.yml [hunk 1]
```diff
@@ -10,6 +10,7 @@ valid_workshop_types:
   - networking
   - f5
   - rhel
+  - security_automation
```
Comment (liquidat) [inline provisioner/group_vars/all.yml]: Small change, but I'd suggest to remove the additional "automation" - all workshops are about automation, so to my understanding the keyword was just differentiating the different use cases.

File: provisioner/provision_lab.yml [hunk 2]
```diff
@@ -166,6 +166,38 @@
...
+  roles:
+    - role: splunk_enterprise
+      when:
+        - workshop_type == "security_automation"
+
+- name: INSTALL AND CONFIGURE SNORT
+  hosts: snort
+  become: true
+  vars:
+    ids_config_snort_version: '2.9.13'
```
Comment (liquidat) [inline provisioner/provision_lab.yml]: How about putting those variables in additional group_vars? Best practices recommend to not put variables into playbooks IIRC.

File: provisioner/provision_lab.yml [hunk 3]
```diff
@@ -166,6 +166,38 @@
...
+    ids_install_provider: "snort"
+    ids_install_normalize_logs: True
+    ids_install_snort_user: root
+    ids_install_snort_group: root
+  tasks:
+    - block:
+      - name: setup epel for snort ecosystem rule lifecycling
+        import_role:
+          name: "geerlingguy.repo-epel"
+      - name: import ids_install role
```
Comment (liquidat) [inline provisioner/provision_lab.yml]: Both IDS roles are commited to the repo here as well. Any reason why you do not use roles/requirements.yml?

File: provisioner/roles/manage_ec2_instances/defaults/main.yml [hunk 4]
```diff
@@ -70,3 +69,19 @@ ec2_info:
...
+  splunk_enterprise:
+    owners: 309956199498
+    size: t2.micro
+    os_type: linux
+    disk_space: 10
+    architecture: x86_64
+    filter: 'RHEL-7.6_HVM-20190515-x86_64-0-Access2-GP2'
+    username: ec2-user
+  checkpoint_firewall:
+    owners: 309956199498
```
Comment (liquidat) [inline provisioner/roles/manage_ec2_instances/defaults/main.yml]: For Checkpoint, please change to `679593333241`.

File: provisioner/roles/manage_ec2_instances/defaults/main.yml [hunk 5]
```diff
@@ -70,3 +69,19 @@ ec2_info:
...
+    size: t2.micro
+    os_type: linux
+    disk_space: 10
+    architecture: x86_64
+    filter: 'RHEL-7.6_HVM-20190515-x86_64-0-Access2-GP2' #FIXME - needs to be something checkpoint
```
Comment (liquidat) [inline provisioner/roles/manage_ec2_instances/defaults/main.yml]: For Checkpoint CloudGuard IaaS please change to `Check Point CloudGuard IaaS*PAYG-MGMT*R80.20*`

File: provisioner/roles/manage_ec2_instances/tasks/addhost_security_automation.yml [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+- name: grab facts for checkpoint node (SECURITY AUTOMATION MODE)
+  ec2_instance_facts:
+    region: "{{ ec2_region }}"
+    filters:
+      instance-state-name: running
+      "tag:Workshop_checkpoint": "{{ec2_name_prefix}}-checkpoint"
+  register: checkpoint_node_facts
+
+- name: grab facts for rtr4 node (SECURITY AUTOMATION MODE)
```
Comment (liquidat) [inline provisioner/roles/manage_ec2_instances/tasks/addhost_security_automation.yml]: What is this for?

File: provisioner/roles/manage_ec2_instances/tasks/create_inventory.yml [hunk 7]
```diff
@@ -33,3 +33,7 @@
 - name: f5 mode - setup inventory
   include_tasks: "addhost_f5.yml"
   when: workshop_type == 'f5'
+
+- name: Security Automation - setup inventory
```
Comment (liquidat) [inline provisioner/roles/manage_ec2_instances/tasks/create_inventory.yml]: Please rename to `security mode - setup inventory` to follow the naming guideline of all other inventories.

File: provisioner/roles/manage_ec2_instances/tasks/instances_security_automation.yml [hunk 8]
```diff
@@ -1,168 +1,164 @@
...
-    node2_node: "{{ ec2_lab_node_types['node2'] }}"
-    node3_node: "{{ ec2_lab_node_types['node3'] }}"
-  when: not networking
-
-- name: find ami for node1
+- name: find ami for splunk
   ec2_ami_facts:
-    region: "{{ ec2_region }}"
-    owners: "{{ ec2_info[node1_node].owners }}"
+    region: "us-east-1" # I don't know why but that seems to be the only place it's available
```
Comment (liquidat) [inline provisioner/roles/manage_ec2_instances/tasks/instances_security_automation.yml]: If this is the case we need to document that in the example yml for the provisioning.

File: provisioner/group_vars/all.yml (hunk 1 above)
Comment (liquidat) [inline provisioner/group_vars/all.yml]: Also, please create an example file in `workshops/provisioner/sample_workshops` to highlight the different variables and note if anything special needs to be taken into consideration.

File: provisioner/roles/manage_ec2_instances/tasks/instances_security_automation.yml [hunk 9]
```diff
@@ -1,168 +1,164 @@
...
+    owners: "{{ ec2_info['splunk_enterprise']['owners'] }}"
     filters:
-      name: "{{ ec2_info[ansible_node].filter }}"
-      architecture: "{{ ec2_info[node1_node].architecture }}"
+      name: "{{ ec2_info['splunk_enterprise']['filter'] }}"
+      architecture: "{{ ec2_info['splunk_enterprise']['architecture'] }}"
   register: amis
 
-- name: save ami for node1
+- name: save ami for splunk
```
Comment (liquidat) [inline provisioner/roles/manage_ec2_instances/tasks/instances_security_automation.yml]: In case of networking we have outsourced the ami part into `ami_find_networking.yml`. Should we do the same with security, moving those bits out of `instances_security_automation.yml` and into a dedicated `ami_find_security.yml`?

I personally would prefer that since to my eyes it makes it much more clean to work with it. But I wouldn't fight for it if you object. What do you think,  @IPvSean and @maxamillion ?

File: provisioner/roles/manage_ec2_instances/defaults/main.yml [hunk 10]
```diff
@@ -70,3 +69,19 @@ ec2_info:
...
     username: admin
+  splunk_enterprise:
+    owners: 309956199498
+    size: t2.micro
+    os_type: linux
+    disk_space: 10
+    architecture: x86_64
+    filter: 'RHEL-7.6_HVM-20190515-x86_64-0-Access2-GP2'
+    username: ec2-user
+  checkpoint_firewall:
```
Comment (liquidat) [inline provisioner/roles/manage_ec2_instances/defaults/main.yml]: Before the firewall we first need the cloud security management server. The following details are for the management server, I can work on the firewalls as soon as we have this PR merged.

File: provisioner/roles/manage_ec2_instances/tasks/instances_security_automation.yml [hunk 11]
```diff
@@ -1,168 +1,164 @@
...
-  when: node2_output.instance_ids is not none
+    - "{{ snort_output['instance_ids'] }}"
+  when: snort_output['instance_ids'] is not none
 
-- name: find ami for node3
+- name: find ami for checkpoint
   ec2_ami_facts:
     region: "{{ ec2_region }}"
-    owners: "{{ ec2_info[node3_node].owners }}"
+    owners: "{{ ec2_info['checkpoint_firewall']['owners'] }}"
```
Comment (liquidat) [inline provisioner/roles/manage_ec2_instances/tasks/instances_security_automation.yml]: See above - we first need to setup the management server, and then add the firewalls.

File: provisioner/roles/splunk_enterprise/files/lets-encrypt-x3-cross-signed.key [hunk 12]
```diff
@@ -0,0 +1,27 @@
+-----BEGIN CERTIFICATE-----
```
Comment (liquidat) [inline provisioner/roles/splunk_enterprise/files/lets-encrypt-x3-cross-signed.key]: What is this about?

File: provisioner/roles/splunk_enterprise/tasks/main.yml [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+    url: 'https://s3.amazonaws.com/linklight.securityautomation/splunk-7.2.6-c0bf0f679ce9-Linux-x86_64.tgz'
+    dest: /tmp/splunk.tgz
+
+- name: Git clone splunk-ansible
+  git:
+    repo: 'https://github.com/splunk/splunk-ansible'
+    version: '7.2.6'
+    dest: '/tmp/splunk-ansible'
+
+- name: Copy splunk-ansible Extra Vars
```
Comment (liquidat) [inline provisioner/roles/splunk_enterprise/tasks/main.yml]: You are encapsulating an ansible-call here? That looks ugly, why is this necessary? Also, this might make it hard to re-use in Tower.

File: provisioner/group_vars/all.yml (hunk 1 above)
Comment (maxamillion) [reply provisioner/group_vars/all.yml]: works for me

File: provisioner/provision_lab.yml (hunk 2 above)
Comment (maxamillion) [reply provisioner/provision_lab.yml]: I'll move them because that's definitely cleaner, but I'm curious what Best Practices guide you're referencing because there's no mention of it that I'm seeing in the [core docs](https://docs.ansible.com/ansible/latest/user_guide/playbooks_best_practices.html#variables-and-vaults).

File: provisioner/provision_lab.yml (hunk 3 above)
Comment (maxamillion) [reply provisioner/provision_lab.yml]: I don't follow.

File: provisioner/roles/manage_ec2_instances/tasks/addhost_security_automation.yml (hunk 6 above)
Comment (maxamillion) [reply provisioner/roles/manage_ec2_instances/tasks/addhost_security_automation.yml]: that is a left over from something that didn't pan out ... will clean up

File: provisioner/roles/manage_ec2_instances/tasks/instances_security_automation.yml (hunk 8 above)
Comment (maxamillion) [reply provisioner/roles/manage_ec2_instances/tasks/instances_security_automation.yml]: @liquidat which example? the `sample_workshops` or $other?

File: provisioner/roles/splunk_enterprise/files/lets-encrypt-x3-cross-signed.key (hunk 12 above)
Comment (maxamillion) [reply provisioner/roles/splunk_enterprise/files/lets-encrypt-x3-cross-signed.key]: It's needed for the provisioning of a letsencrypt cert, I borrowed it from `./provisioner/roles/aws_dns/files/lets-encrypt-x3-cross-signed.key` along with logic from that role.

File: provisioner/roles/splunk_enterprise/tasks/main.yml (hunk 13 above)
Comment (maxamillion) [reply provisioner/roles/splunk_enterprise/tasks/main.yml]: @liquidat I am, I'm using [splunk-ansible](https://github.com/splunk/splunk-ansible) for the install of splunk but that particular set of playbooks is explicitly designed to be run on the local host of the splunk instance (yes, this is an anti-pattern and I hope to sync with that team to resolve that)

File: provisioner/roles/splunk_enterprise/files/lets-encrypt-x3-cross-signed.key (hunk 12 above)
Comment (liquidat) [reply provisioner/roles/splunk_enterprise/files/lets-encrypt-x3-cross-signed.key]: Ok.

File: provisioner/roles/splunk_enterprise/tasks/main.yml (hunk 13 above)
Comment (liquidat) [reply provisioner/roles/splunk_enterprise/tasks/main.yml]: Ah, good to know, thanks.

File: provisioner/provision_lab.yml (hunk 2 above)
Comment (liquidat) [reply provisioner/provision_lab.yml]: Yeah, sorry for that, I might have dreamed that. So skip this comment.

File: provisioner/provision_lab.yml (hunk 3 above)
Comment (liquidat) [reply provisioner/provision_lab.yml]: What I mean: you added these roles to this repository. Another way would be to just create a [roles/requirements.yml](https://docs.ansible.com/ansible/latest/reference_appendices/galaxy.html#installing-multiple-roles-from-a-file) file to just "link" to the roles and install them on the managed node when needed via `ansible-galaxy` - instead of them being part of the repository.

The advantages are: less clutter in our repo, and it is easier to keep the roles updated.

File: provisioner/roles/manage_ec2_instances/defaults/main.yml [hunk 14]
```diff
@@ -70,3 +69,19 @@ ec2_info:
     architecture: x86_64
     filter: 'F5 Networks Prelicensed Hourly BIGIP-13.1*Good 25MBPS*'
     username: admin
+  splunk_enterprise:
+    owners: 309956199498
+    size: t2.micro
```
Comment (liquidat) [inline provisioner/roles/manage_ec2_instances/defaults/main.yml]: Immediately upon login I got a warning about running out of space. We should pick something larger, for example a c4.4xlarge. See the [Splunk AWS tech brief](https://www.splunk.com/pdfs/technical-briefs/deploying-splunk-enterprise-on-amazon-web-services-technical-brief.pdf): that is of course for production usage but recommends even larger instances.

Comment (maxamillion) [reply provisioner/roles/manage_ec2_instances/defaults/main.yml]: @liquidat +1 - I'm like 90% sure I meant to do that and then just didn't

File: provisioner/roles/manage_ec2_instances/defaults/main.yml [hunk 15]
```diff
@@ -70,3 +69,19 @@ ec2_info:
     architecture: x86_64
     filter: 'F5 Networks Prelicensed Hourly BIGIP-13.1*Good 25MBPS*'
     username: admin
+  splunk_enterprise:
+    owners: 309956199498
+    size: c4.4xlarge
+    os_type: linux
+    disk_space: 10
```
Comment (liquidat) [inline provisioner/roles/manage_ec2_instances/defaults/main.yml]: Please increase this. With 10GB we run into warnings directly after deployment, and run out of storage quickly as soon as the add-on is installed. To be save I'd suggest something around 25 GB, or even 50GB?

File: provisioner/roles/manage_ec2_instances/tasks/instances_security.yml [hunk 16]
```diff
@@ -0,0 +1,164 @@
+- name: find ami for splunk
```
Comment (liquidat) [inline provisioner/roles/manage_ec2_instances/tasks/instances_security.yml]: Should we not put this find-ami-code into its own task? Just like we did it with the networking find-ami stuff?
@IPvSean, comments?

Comment (IPvSean) [reply provisioner/roles/manage_ec2_instances/tasks/instances_security.yml]: I am flexible in this... I try to divide up the Ansible Playbook into 150 or less lines so they are human readable.  This Ansible Playbook is still not that long... but it would be nice to match?  I feel like the way I organize this is getting confusing and I would be OK with rethinking how we organize this.  @maxamillion do you have an opinion?

File: provisioner/roles/manage_ec2_instances/defaults/main.yml (hunk 15 above)
Comment (IPvSean) [reply provisioner/roles/manage_ec2_instances/defaults/main.yml]: I will defer to Roland, we are not getting beat up yet by pricing models, etc so I would encourage you to just increase it, so its reliable

File: provisioner/roles/manage_ec2_instances/tasks/instances_security_automation.yml (hunk 9 above)
Comment (IPvSean) [reply provisioner/roles/manage_ec2_instances/tasks/instances_security_automation.yml]: yeah I tried to break it up to make it cleaner

**Comment by IPvSean** [comment]
sorry for delay in reviewing, i added some comments above, mostly I feel like... maybe I don't understand, but can we just dynamically make requirements with all these roles (looks like 4?) and load them each time?  I feel like by including them here we effectively have a fork of their role? @liquidat

**Comment by liquidat** [comment]
> sorry for delay in reviewing, i added some comments above, mostly I feel like... maybe I don't understand, but can we just dynamically make requirements with all these roles (looks like 4?) and load them each time? I feel like by including them here we effectively have a fork of their role? @liquidat

You mean the geerlingguy and similar roles? I would also like to dynamically include them via roles/requirements.yml.

File: provisioner/provision_lab.yml (hunk 3 above)
Comment (maxamillion) [reply provisioner/provision_lab.yml]: @liquidat I follow, I'll do that.

File: provisioner/roles/manage_ec2_instances/tasks/instances_security.yml (hunk 16 above)
Comment (maxamillion) [reply provisioner/roles/manage_ec2_instances/tasks/instances_security.yml]: it's possible the "find AMI" tasks might need to live in their own task file at this point and be included at the appropriate point in the `main.yml` for clarity ... this file is approaching unwieldy status, it's definitely not there yet but as we grow the workshop more I could see us getting there without too much effort

**Comment by IPvSean** [comment]
this is awesome

THIS IS AWESOME
